### PR TITLE
mvcc: skip Value bytes from BoltDB for KeysOnly range queries (-77% latency, -97% heap at 10 KB/key)

### LINKED ISSUE #20386: Optimize Range WithKeysOnly to avoid loading values in memory
### What would you like to be added?

Kubernetes 1.34 will start using [Range WithKeysOnly](https://github.com/kubernetes/kubernetes/pull/132355). Main reason to use this method is to reduce the overhead of loading values in memory when they are not needed. While using WithKeysOnly avoids transferring them to client, it doesn't help with etcd memory

Noticed that it loads values in memory just to throw them out little later: https://github.com/etcd-io/etcd/blob/810d69619e33aa09bfdc0449dc65ceb34e589ab6/server/etcdserver/txn/range.go#L150-L152

TODO:
* Add thorough range tests we don't break it.
* Identify places that use value in etcd server. For example when sorting by value 
* Implement support for KeysOnly method for keyindex
* Use KeysOnly method when we don't need value, like when we don't sort by it.
* Benchmark to show improvement

Please note this task is complicated and not good for first contributor.

### Why is this needed?

Align with user expectation that using WithKeysOnly is more efficient on server.

### PR DESCRIPTION
**Pull request by notandruu**
Fixes #20386

## Root Cause

When `RangeRequest.KeysOnly=true`, etcd still loaded the full serialized
`KeyValue` proto from BoltDB — including potentially multi-kilobyte `Value`
bytes — via `proto.Unmarshal`, then immediately discarded the value in
`asembleRangeResponse`:

```go
// server/etcdserver/txn/range.go (before)
if r.KeysOnly {
    rr.KVs[i].Value = nil   // ← bytes were already allocated and copied from BoltDB
}
```

Kubernetes 1.34 uses `Range WithKeysOnly` for watch cache warming and list
operations. For a cluster with 10 000 secrets (10 KB each), a single
`LIST` with `KeysOnly=true` caused ~100 MB of transient heap allocations
that were discarded immediately.

## Fix

Three-layer change:

1. **`server/storage/mvcc/kv.go`** — add `KeysOnly bool` to `RangeOptions`.

2. **`server/etcdserver/txn/range.go`** — set `ro.KeysOnly = r.KeysOnly && r.SortTarget != VALUE`
   in `executeRange`. When the request sorts by `VALUE`, values are still
   required for the sort step; all other `KeysOnly` requests can skip them.

3. **`server/storage/mvcc/kvstore_txn.go`** — add `unmarshalKVSkipValue`, a hand-rolled
   `protowire` decoder that reads all `KeyValue` fields (Key, CreateRevision,
   ModRevision, Version, Lease) and skips proto field 5 (`Value`) entirely,
   never touching the backing buffer for those bytes. When `ro.KeysOnly`,
   `rangeKeys` calls this function instead of `proto.Unmarshal`.

The existing `if r.KeysOnly { rr.KVs[i].Value = nil }` in
`asembleRangeResponse` is kept as a safety net for the `SortTarget == VALUE`
case (values loaded → sorted → stripped).

## Benchmark Results

Measured on Apple M3 Max, go1.26.3, darwin/arm64. 100 keys per range, each
key stored with a value of the given size.

```
goos: darwin
goarch: arm64
cpu: Apple M3 Max

BenchmarkRangeKeysOnly1KB_WithValue-16    111879    58458 ns/op    136057 B/op    719 allocs/op
BenchmarkRangeKeysOnly1KB_KeysOnly-16     168064    34670 ns/op     33656 B/op    619 allocs/op
                                                     -41% latency    -75% heap

BenchmarkRangeKeysOnly10KB_WithValue-16    40623   142281 ns/op   1057660 B/op    719 allocs/op
BenchmarkRangeKeysOnly10KB_KeysOnly-16    180662    32728 ns/op     33656 B/op    619 allocs/op
                                                     -77% latency    -97% heap
```

Heap savings scale linearly with value size because the Value allocation is
eliminated entirely. The 100 B of alloc difference (719 → 619 per 100-key
range) is the removal of per-KV Value slice headers from the allocator.

## Tests Added

- `TestUnmarshalKVSkipValue` — unit test for the protowire decoder: verifies
  all non-Value fields are correctly decoded and Value is always nil.
- `TestUnmarshalKVSkipValueEmptyValue` — edge case: tombstone records with no
  value field decode cleanly.
- `TestRangeKeysOnlyDoesNotLoadValues` — end-to-end: stores keys with values,
  ranges with `KeysOnly=true` and asserts `Value == nil` while all metadata
  fields (Key, CreateRevision, ModRevision, Version) are correctly populated;
  also verifies the normal (non-KeysOnly) path still returns values.
- `BenchmarkRangeKeysOnly*` — four benchmarks (1 KB / 10 KB × with/without
  KeysOnly) to anchor future regressions.

### CODE REVIEW

**Comment by serathius** [comment]
cc @nwnt

**Comment by notandruu** [comment]
Could a maintainer run `/ok-to-test` to unblock CI? Happy to address any feedback.

**Comment by serathius** [comment]
This is something that @nwnt was already working on and we discussed a design that was different than presented here.

File: server/storage/mvcc/kvstore_txn.go [hunk 1]
```diff
@@ -122,18 +123,96 @@ func (tr *storeTxnCommon) rangeKeys(ctx context.Context, key, end []byte, curRev
...
 
+// unmarshalKVSkipValue decodes a serialized mvccpb.KeyValue proto message into
+// kv without allocating memory for the Value field. It is used when the caller
+// only needs key metadata (key name, revisions, version, lease) and the Value
+// bytes would be immediately discarded — for example, in KeysOnly range queries.
+//
+// Fields are decoded by hand using protowire so that the Value bytes (proto
+// field 5) are never copied from the backend buffer into Go heap memory. All
+// other fields are decoded as normal. Unknown future fields are skipped safely.
+func unmarshalKVSkipValue(b []byte, kv *mvccpb.KeyValue) error {
```
Comment (serathius) [inline server/storage/mvcc/kvstore_txn.go:155]: Don't want to maintain custom protobuf decoder.

**Comment by serathius** [comment]
Fixed via https://github.com/etcd-io/etcd/pull/21791
